### ISSUE TITLE
[RFC] Cost-based coloration scheme and its problem

### ISSUE THREAD

**Issue by kimikage** [timholy/FlameGraphs.jl#27]
I think the coloring based on the costs (i.e. sample or time counts) is a natural visualization scheme. 
However, the current API is not suitable for the scheme.

The following is a hack that "abuses" the current API.
```julia
using FlameGraphs, Colors, FixedPointNumbers
import FlameGraphs: NodeData

struct CostBasedColors{mode}
    fc::FlameColors
    CostBasedColors{mode}(fc::FlameColors) where {mode} = new{mode}(fc)
end

function (::Type{T})(oddcolors::AbstractVector{<:Color}, 
                     evencolors::AbstractVector{<:Color}=oddcolors; 
                     kwargs...) where T <: CostBasedColors
    length(oddcolors) == length(evencolors) || error("oddcolors and evencolors must have the same length.")
    fc = FlameColors(0; kwargs...)
    append!(fc.colors, RGB{N0f8}.(oddcolors))
    append!(fc.colors, RGB{N0f8}.(evencolors))
    T(fc)
end

(colors::CostBasedColors)(s::Symbol) = colors.fc(s)
```

### Total-cost-based

```julia
const TotalCostBasedColors = CostBasedColors{:total}

function (colors::TotalCostBasedColors)(nextidx, j::Integer, data::NodeData)
    fc = colors.fc
    n = length(fc.colors) ÷ 2
    if j == 1
        nextidx[1] = 0
        c = fc(nextidx, j, data)
        nextidx[1] = length(data.span)
    else
        maxspan = nextidx[1]
        total = length(data.span)
        nextidx[j] = round(Int, total / maxspan * (n - 1)) + 1
        c = fc(nextidx, j, data)
    end
    return c
end
```

```julia
tcolor = TotalCostBasedColors(colormap("Blues", 10), colormap("Greens", 10), colorbg=colorant"silver");
flamepixels(tcolor, flamegraph())|>rotl90
```
![total](https://user-images.githubusercontent.com/12679384/81497043-df0cf580-92f6-11ea-87bd-3f97a2f0da0f.png)

This concept is simple but the results are not so useful because the total cost is already visualized as the width.

### Relative-cost-based
This scheme is somewhat tricky. It is based on the "complement" of the cost rate of its own frame relative to the parent frame.
```julia
const RelativeCostBasedColors  = CostBasedColors{:relative}

function (colors::RelativeCostBasedColors)(nextidx, j::Integer, data::NodeData)
    fc = colors.fc
    n = length(fc.colors) ÷ 2
    if j == 1
        nextidx[1] = 1
        c = fc(nextidx, j, data)
    else
        parent = nextidx[j - 1]
        total = length(data.span)
        nextidx[j] = round(Int, (parent - total) / parent * (n - 1)) + 1
        c = fc(nextidx, j, data)
    end
    nextidx[j] = length(data.span)
    return c
end
```

```julia
rcolor = RelativeCostBasedColors(colormap("Blues", 10), colorbg=colorant"silver");
flamepixels(rcolor, flamegraph())|>rotl90
```
![relative](https://user-images.githubusercontent.com/12679384/81497339-c43b8080-92f8-11ea-8cb5-d0ea947423e8.png)

In short, this is a substitute of the "self-cost-based" scheme.

### Problem
The problem with the cost-based scheme is that the tree information cannot be obtained from the coloring API. However, I don't want to break the compatibility. :confused:

**Comment by kimikage** [comment]
This is a separate request, but there are similar issues with the `filter` option. For example, the frames with a small self (own) cost are not important, so I want to collapse them. However, `filter` does not support that.

**Comment by kimikage** [comment]
Perhaps for the time being, it seems better to extend it on the front ends.

**Comment by timholy** [comment]
Since we already have one axis for cost, I don't personally think it makes sense to duplicate this information with color.

**Comment by kimikage** [comment]
I don't think the "self-cost" is so clear. In other words, there is too much noise with zero self-costs.

**Comment by timholy** [comment]
Interesting. Can you show an example? Let's document the problem first before getting to a solution.

As I'm sure you know, with any sampling profiler you expect `O(sqrt(n))` noise.

**Comment by kimikage** [comment]
I'm sorry I didn't make it clear enough. I meant the visual noise when analyzing the result, not the noise by sampling.
I think the whitish pixels in the second figure are often a waste of space.

**Comment by timholy** [comment]
Ah, you mean you're trying to use color to direct attention to the lines that do the actual work? (Essentially like the "overhead" entries in profile text dumps?) I'm more sympathetic now. However, I wonder in practice if you'd still need to vary the coloration of the callers. Rarely do I find it useful that the actual work is being done by `+` in `float.jl`, except as reassurance that I don't have something obviously broken about the performance in my code. The interesting details are typically in who ends up calling `+` and whether the number of times can be reduced.
